### ISSUE TITLE
String Similarity Library/ Algorithm

### ISSUE THREAD

**Issue by teskuteyi** [rapidsai/cudf#17184]
I am trying to be able to run this anonymized script using cuDF or cuDF.pandas with GPU acceleration.
currently runs for over 5 hours for 2Million rows of data.

```python
import pandas as pd
from rapidfuzz import fuzz, process

def fuzzy_merge(df1, df2, key1, key2, threshold=95, limit=1):
    """
    Perform a fuzzy merge and return both matching and unmatching DataFrames.
    """
    s = df2[key2].tolist()
    m = df1[key1].apply(lambda x: process.extract(x, s, scorer=fuzz.partial_ratio, limit=limit))
    df1['matches'] = m

    m2 = pd.DataFrame(df1['matches'].tolist(), index=df1.index)
    df1['match'] = m2[0].apply(lambda x: x[0] if x[1] >= threshold else None)
    
    # Drop the 'matches' column as it's no longer needed
    df1 = df1.drop(columns=['matches'], axis=1)
    
    # Merge based on the 'match' column to include all rows from df1
    df_final = pd.merge(df1, df2, left_on='match', right_on=key2, how='left')
    
    # Identify matching and unmatching rows
    matching_rows = df_final.dropna(subset=[key2])  # Rows where the merge found a match
    unmatching_rows = df_final[df_final[key2].isna()]  # Rows without a match

    return matching_rows, unmatching_rows.drop(columns=df2.columns, axis=1)

# File paths
input_file1 = 'csvfile1.csv'
input_file2 = 'csvfile2.csv'
output_file_matched = 'csvfile3.csv'
output_file_unmatched = 'csvfile4.csv'

# Specify the key columns in both DataFrames
key1 = 'column_name_in_df1'  # Replace with actual column name from df1
key2 = 'column_name_in_df2'  # Replace with actual column name from df2

try:
    # Load your datasets
    df1 = pd.read_csv(input_file1)
    df2 = pd.read_csv(input_file2)

    # Perform the fuzzy merge and separate the results
    matched_df, unmatched_df = fuzzy_merge(df1, df2, key1, key2, threshold=95)

    # Save the matching and unmatching rows to their respective CSV files
    matched_df.to_csv(output_file_matched, index=False)
    unmatched_df.to_csv(output_file_unmatched, index=False)

    print(f"Matched rows saved to {output_file_matched}")
    print(f"Unmatched rows saved to {output_file_unmatched}")

```

**Comment by bdice** [comment]
With `cudf.pandas`, the line `df1[key1].apply(lambda x: process.extract(x, s, scorer=fuzz.partial_ratio, limit=limit))` will not be able to run on GPU because of limitations on what is supported in [user-defined `apply` functions](https://docs.rapids.ai/api/cudf/stable/user_guide/guide-to-udfs/#caveats). However, I think we have the tools needed to accelerate this computation.

From StackOverflow, it sounds like the partial ratio is computed using Levenshtein distances: https://stackoverflow.com/questions/53755558/need-more-understanding-on-python-fuzz-partial-ratio

cuDF has a function for `edit_distance` which computes the Levenshtein distance. https://docs.rapids.ai/api/cudf/stable/user_guide/api_docs/api/cudf.core.column.string.stringmethods.edit_distance/

Can you try using that and let us know?

**Comment by davidwendt** [comment]
There is also `jaccard_index` function. Looks like it is not documented but should be.
https://github.com/rapidsai/cudf/blob/4c04b7c8790263dc68c5753609f3cb867806359f/python/cudf/cudf/core/column/string.py#L5463

Here is the unit test which can serve as an example usage
https://github.com/rapidsai/cudf/blob/4c04b7c8790263dc68c5753609f3cb867806359f/python/cudf/cudf/tests/text/test_text_methods.py#L1009

**Comment by wence-** [comment]
@teskuteyi is there anything else you need here? For example, guidance on how to use the suggested features?

**Comment by davidwendt** [comment]
@VibhuJawa may also have recommendations on this

**Comment by VibhuJawa** [comment]
Thanks for raising this! As David mentioned, you can likely use the Jaccard index to compute similarity between string columns. 

We use a similar technique at terabyte-scale in [NeMo Curator](https://github.com/NVIDIA/NeMo-Curator/blob/20fd23709f3ae543d5a8e19b5ccbe48b91ef74dc/nemo_curator/modules/fuzzy_dedup/jaccardsimilarity.py#L138-L154) for deduplication tasks.

Here’s a minimal working example that might help you get started with cudf:

```python3
import cudf

def jaccard_fuzzy_merge(
    df_left: cudf.DataFrame,
    df_right: cudf.DataFrame,
    left_key: str,
    right_key: str,
    width: int = 5,
    threshold: float = 0.95,
):
    """
    GPU‑only Jaccard‑similarity merge.

    Returns
    -------
    matched_df   – row_id | left_val | right_val | jaccard
    unmatched_df – all original columns from `df_left` that found no match ≥ threshold
    """
    # ────────────────────────────────────── prep left / right columns
    lcol, rcol = "left_val", "right_val"
    left  = df_left[[left_key]].rename(columns={left_key: lcol}).reset_index()
    left  = left.rename(columns={"index": "row_id"})            # preserve original row #
    right = df_right[[right_key]].rename(columns={right_key: rcol})

    # ───────────────────────────── Cartesian product fully on‑GPU
    pairs = (
        left.assign(_tmp=1)
            .merge(right.assign(_tmp=1), on="_tmp")
            .drop(columns="_tmp")
    )

    # ─────────────────────────────── Jaccard similarity on strings
    pairs["jaccard"] = pairs[lcol].str.jaccard_index(pairs[rcol], width)

    # ───────────── keep the best (highest‑jaccard) candidate per left row
    pairs = (
        pairs.sort_values(["row_id", "jaccard"], ascending=[True, False])
             .groupby("row_id", as_index=False)
             .first()
    )
    # ───────────────────────── split matched vs unmatched
    matched_mask  = pairs["jaccard"] >= threshold
    matched_pairs = pairs.loc[matched_mask, ["row_id", lcol, rcol, "jaccard"]].reset_index(drop=True)

    # use a MERGE (still GPU) instead of .take()/iloc for unmatched rows
    unmatched_ids  = pairs.loc[~matched_mask, "row_id"].astype("int32")
    unmatched_df   = (
        df_left.reset_index()                                   # add original row_id
               .rename(columns={"index": "row_id"})
               .merge(unmatched_ids, on="row_id", how="inner")
               .drop(columns="row_id")
               .sort_index(ignore_index=True)                   # optional tidy‑up
    )

    return matched_pairs, unmatched_df
```

```
df1 = cudf.DataFrame({"id_1": [1, 2], "name": ["NVIDIA Corp.", "Rapids-AI"]})
df2 = cudf.DataFrame({"id_2": [1, 2], "legal": ["NVIDIA Corporation", "Rapid‑AI"]})

matched_pairs, unmatched_df = jaccard_fuzzy_merge(df1, df2, "name", "legal", threshold=0.4)
print("-"*10+"Matched"+"-"*10)
print(matched_pairs)
print("-"*10+"Unmatched"+"-"*10)
print(unmatched)
```


```
----------Matched----------
   row_id      left_val           right_val   jaccard
0       0  NVIDIA Corp.  NVIDIA Corporation  0.466667
----------Unmatched----------
   id_1          name
0     1  NVIDIA Corp.
1     2     Rapids-AI
```


Tweak threshold, case‑folding, punctuation stripping, etc., to suit your data.

**Scaling tips**
- MinHash + Jaccard — generate MinHashes first, bucket similar strings, then run the exact Jaccard step only within buckets (this is how NeMo Curator stays tractable at TB scale).

- Dask‑cuDF — for true N×N comparisons, partition the Cartesian product with Dask so the work is chunked across GPUs.

Helpful notebook: single‑GPU fuzzy dedup : https://github.com/NVIDIA/NeMo-Curator/blob/20fd23709f3ae543d5a8e19b5ccbe48b91ef74dc/tutorials/single_node_tutorial/single_gpu_tutorial.ipynb#L36

Hope this unblocks you—let us know how it goes!

**Comment by vyasr** [comment]
Thanks Vibhu! Closing for now, but please feel free to ask if you have more questions.

**Comment by teskuteyi** [comment]
Thank you @VibhuJawa , working on your suggestions but running into some  memory errors when it comes to the cartesian product even with Dask-cuDF

**Comment by VibhuJawa** [comment]
>Thank you @VibhuJawa , working on your suggestions but running into some memory errors when it comes to the cartesian product even with Dask-cuDF


I think this is an expansive merge.  I will suggest doing something like this to make it suceed at scale.  Worth a shot . 


```python3
#!/usr/bin/env python
"""
Minimal Cartesian merge with dask‑cuDF
--------------------------------------
• Starts a LocalCUDACluster only when run as a script.
• Repartitions each input to ≤100 MB chunks.
• Performs the Cartesian product via a dummy “_tmp” column.
"""

from dask_cuda import LocalCUDACluster
from dask.distributed import Client
import cudf
import dask_cudf


def cartesian_merge(
    left: "dask_cudf.DataFrame",
    right: "dask_cudf.DataFrame",
    partition_size: str = "100MB",
) -> "dask_cudf.DataFrame":
    """Return the Cartesian product (|left| × |right|)."""
    left = left.repartition(partition_size=partition_size)
    right = right.repartition(partition_size=partition_size)

    return (
        left.assign(_tmp=1)
            .merge(right.assign(_tmp=1), on="_tmp")
            .drop("_tmp", axis=1)
    )


if __name__ == "__main__":
    # 1. spin up one worker per visible GPU
    cluster = LocalCUDACluster()
    client = Client(cluster)
   
   # 2. toy inputs (swap in your real cuDFs)
   left_gdf = cudf.DataFrame({"a": cudf.Series.arange(5)})
   right_gdf = cudf.DataFrame({"b": cudf.Series.arange(3)})

   # 3. distribute to Dask (one partition/GPU as a starting point)
   nparts = len(cluster.worker_spec)
   dleft = dask_cudf.from_cudf(left_gdf, npartitions=nparts)
   dright = dask_cudf.from_cudf(right_gdf, npartitions=nparts)

   # 4. Cartesian merge ➜ kick work to the cluster, then inspect
   pairs = cartesian_merge(dleft, dright)
   print("Sample result:\n", pairs.head().compute())

```

**Comment by teskuteyi** [comment]
Thank you very much @VibhuJawa , well appreciated. All working now!
